Orange/widgets/utils/owlearnerwidget.py

```diff
@@ -1,26 +1,207 @@
+        self.error(self.DATA_ERROR_ID)
+        self.data = data
+        if data is not None and data.domain.class_var is None:
+            self.error(self.DATA_ERROR_ID, "Data has no target variable")
+            self.data = None
 
-class OWProvidesLearner(_OWAcceptsPreprocessor):
-    """
-    Base class for all classification / regression learner-providing widgets
-    that extend it.
-    """
+        self.update_model()
+
+    def apply(self):
+        """Applies leaner and sends new model.
+        """
+        self.update_learner()
+        self.update_model()
+
+    @abstract
+    def create_learner(self):
+        """Creates a learner with current configuration.
+
+        Returns:
+            Leaner: an instance of Orange.base.learner subclass.
+        """
+        raise NotImplementedError()
+
+    def update_learner(self):
+        self.learner = self.create_learner()
+        self.send("Learner", self.learner)
+        self.outdated_settings = False
+        self.warning(self.OUTDATED_LEARNER_WARNING_ID)
+
+    def update_model(self):
+        self.model = None
+        self.good_data = False
+        if self.data is not None:
+            self.error(self.DATA_ERROR_ID)
+            if not self.learner.check_learner_adequacy(self.data.domain):
+                self.error(self.DATA_ERROR_ID, self.learner.learner_adequacy_err_msg)
+            elif len(np.unique(self.data.Y)) < 2:
+                self.error(self.DATA_ERROR_ID,
+                           "Data contains single target value. "
+                           "There is nothing to learn.")
+            else:
+                self.model = self.learner(self.data)
+                self.model.name = self.learner_name
+                self.model.instances = self.data
+                self.good_data = True
+
+        self.send(self.OUTPUT_MODEL_NAME, self.model)
+
+    def settings_changed(self, *args, **kwargs):
+        self.outdated_settings = True
+        self.warning(self.OUTDATED_LEARNER_WARNING_ID,
+                     "Press Apply to submit changes.")
+
+    def get_model_parameters(self):
+        """Creates an `OrderedDict` or a sequence of pairs with current model
+        configuration.
+
+        Returns:
+            OrderedDict or List: (option, value) pairs or dict
+        """
+        return None
+
+    def send_report(self):
+        self.report_items((("Name", self.learner_name),))
+
+        model_parameters = self.get_model_parameters()
+        if model_parameters:
+            self.report_items("Model parameters", model_parameters)
+
+        if self.data:
+            self.report_data("Data", self.data)
+
+    # GUI
+    def add_learner_name_widget(self):
+        gui.lineEdit(self.controlArea, self, 'learner_name', box='Name',
+                     tooltip='The name will identify this model in other widgets')
+
+    def add_main_layout(self):
+        """Creates layout with the learner configuration widgets.
+        """
```

Comment:
Please move these meant-to-be-overridden methods to the top. Perhaps also be more verbose:
> Creates layout with the learner configuration widgets.

> Override this method for laying out any learner-specific parameter controls.
> See ``setup_layout()`` for execution order.